cscs-checks/microbenchmarks/gpu/gpu_burn/src/gpu_burn.cu

```diff
@@ -39,553 +42,369 @@
-		     blockIdx.x*blockDim.x + threadIdx.x; // X
-
-    int myFaulty = 0;
-    for (size_t i = 1; i < iters; ++i)
-        if (fabsf(C[myIndex] - C[myIndex + i*iterStep]) > EPSILON)
-            myFaulty++;
-
-    atomicAdd(faultyElems, myFaulty);
+namespace kernels
+{
+  template<class T>
+  __global__ void compare(T *C, int *numberOfErrors, size_t iters) {
+      size_t iterStep = blockDim.x*blockDim.y*gridDim.x*gridDim.y;
+      size_t myIndex = (blockIdx.y*blockDim.y + threadIdx.y)* // Y
+                        gridDim.x*blockDim.x + // W
+                        blockIdx.x*blockDim.x + threadIdx.x; // X
+
+      int localErrors = 0;
+      for (size_t i = 1; i < iters; ++i)
+          if (fabs(C[myIndex] - C[myIndex + i*iterStep]) > EPSILOND)
+              localErrors++;
+
+      atomicAdd(numberOfErrors, localErrors);
+  }
 }
 
-extern "C" __global__ void compareD(double *C, int *faultyElems, size_t iters) {
-    size_t iterStep = blockDim.x*blockDim.y*gridDim.x*gridDim.y;
-    size_t myIndex = (blockIdx.y*blockDim.y + threadIdx.y)* // Y
-                     gridDim.x*blockDim.x + // W
-		     blockIdx.x*blockDim.x + threadIdx.x; // X
+template <class T> class GemmTest
+{
+private:
+  int deviceId;
+
+  // SMI handle to do system queries.
+  Smi * smi_handle;
+
+  // Iterations per call to this->compute
+  size_t iters;
+
+  long long int totalErrors;
+
+  // Work arrays
+  T * d_C;
+  T * d_A;
+  T * d_B;
+
+  int * d_numberOfErrors;
+
+  XblasHandle_t d_blas;
+  static const int g_blockSize = 16;
+
+public:
+  GemmTest(int id, Smi * smi_hand) : deviceId(id), smi_handle(smi_hand)
+  {
+      // Set the device and pin thread to CPU with
+      XSetDevice(deviceId);
+      smi_handle->setCpuAffinity(deviceId);
+
+      // Create blas plan
+      XblasCreate(&d_blas);
+
+      totalErrors = 0;
+  }
+  ~GemmTest()
+  {
+      XFree(d_C);
+      XFree(d_A);
+      XFree(d_B);
+      XFree(d_numberOfErrors);
+      XblasDestroy(d_blas);
+      XDeviceSynchronize();
+  }
+
+  unsigned long long int getErrors()
+  {
+      unsigned long long int tempErrs = totalErrors;
+      totalErrors = 0;
+      return tempErrs;
+  }
+
+  size_t getIters()
+  {
+      return iters;
+  }
+
+  size_t availMemory()
+  {
+      size_t freeMem;
+      smi_handle->getDeviceAvailMemorySize(deviceId, &freeMem);
+      return freeMem;
+  }
+
+  void initBuffers(T * h_A, T * h_B)
+  {
+      size_t useBytes = (size_t)((double)availMemory()*USEMEM);
+      size_t d_resultSize = sizeof(T)*SIZE*SIZE;
+      iters = (useBytes - 2*d_resultSize)/d_resultSize; // We remove A and B sizes
+      XMalloc((void**)&d_C, iters*d_resultSize);
+      XMalloc((void**)&d_A, d_resultSize);
+      XMalloc((void**)&d_B, d_resultSize);
+      XMalloc((void**)&d_numberOfErrors, sizeof(int));
+
+      // Populating matrices A and B
+      XMemcpy(d_A, h_A, d_resultSize, XMemcpyHostToDevice);
+      XMemcpy(d_B, h_B, d_resultSize, XMemcpyHostToDevice);
+  }
+
+  void compute()
```

Comment:
I would only define this in the specializations and let it fail in compilation, or you could use a static assert.

Reply:
I've gone for `void compute() = delete;`

Reply:
Yes, better.